emcc.py

```diff
@@ -2451,23 +2496,90 @@ def generate_html(target, options, js_target, target_basename,
         URL.revokeObjectURL(script.src);
       };
       document.body.appendChild(script);
-    };
-    codeXHR.send(null);
-''' % (os.path.basename(asm_target), '\n'.join(asm_mods), script.inline)
+    }
+    if (fileBytes) {
+      onCodeLoad(intArrayToString(fileBytes));
+    } else {
+      codeXHR = new XMLHttpRequest();
+      codeXHR.open('GET', filename, true);
+      codeXHR.onload = function() {
+        onCodeLoad(codeXHR.responseText);
+      };
+      codeXHR.send(null);
+    }
+''' % (shared.JS.get_subresource_location(asm_target), '\n'.join(asm_mods), script.inline)
 
   if shared.Settings.BINARYEN and not shared.Settings.BINARYEN_ASYNC_COMPILATION:
     # We need to load the wasm file before anything else, it has to be synchronously ready TODO: optimize
     script.un_src()
     script.inline = '''
-          var wasmXHR = new XMLHttpRequest();
-          wasmXHR.open('GET', '%s', true);
-          wasmXHR.responseType = 'arraybuffer';
-          wasmXHR.onload = function() {
-            Module.wasmBinary = wasmXHR.response;
+          var filename = '%s';
+          var fileBytes = tryParseAsDataURI(filename);
+          function onWasmLoad() {
 %s
-          };
-          wasmXHR.send(null);
-''' % (os.path.basename(wasm_binary_target), script.inline)
+          }
+          if (fileBytes) {
+            Module.wasmBinary = fileBytes.buffer;
+            onWasmLoad();
+          } else {
+            var wasmXHR = new XMLHttpRequest();
+            wasmXHR.open('GET', filename, true);
+            wasmXHR.responseType = 'arraybuffer';
+            wasmXHR.onload = function() {
+              Module.wasmBinary = wasmXHR.response;
+              onWasmLoad();
+            };
+            wasmXHR.send(null);
+          }
+''' % (shared.JS.get_subresource_location(wasm_binary_target), script.inline)
+
+  if script.inline:
+    # add functions copied from preamble.js
+    script.inline += '''
+      function intArrayToString(array) {
```

Comment:
For the main JS, you can add code in `src/jsifier.js`, see how it adds the code for DETERMINISTIC for example.

For the html, you can just have emcc read that file and use that to append to `script.inline` etc